### PR TITLE
[Steward] Circles USDC CCTP v2 Bridge

### LINKED ISSUE #10: [Steward] Circles USDC CCTP v2 Bridge

### PR DESCRIPTION
**Pull request by jubeira**
Simplified CCTP bridge, based on https://github.com/aave-dao/aave-helpers/pull/667.

Changes from the original version:
- Bridging limited from L2s where the steward is deployed to mainnet only, with target domain and receiver hardcoded at construction time.
  - Functions related to address allowlisting, non evm transfers, etc removed.
- Extra checks added for input parameters
- Reset allowance after bridge

Same target networks as the original PR: Arbitrum, Base, Optimism, Polygon.

### CODE REVIEW

File: src/bridges/cctp/CctpBridgeSteward.sol [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  address public immutable TOKEN_MESSENGER;
+
+  /// @inheritdoc ICctpBridgeSteward
+  address public immutable USDC;
+
+  /// @inheritdoc ICctpBridgeSteward
+  address public immutable COLLECTOR;
+
+  /// @inheritdoc ICctpBridgeSteward
+  address public immutable RECEIVER;
```
Comment (efecarranza) [inline src/bridges/cctp/CctpBridgeSteward.sol]: Maybe rename all of these receiver related to "Mainnet Collector"?

File: src/bridges/cctp/CctpBridgeSteward.sol [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    RECEIVER = receiver;
+
+    address localMessageTransmitter = ITokenMessengerV2(tokenMessenger).localMessageTransmitter();
+    LOCAL_DOMAIN = IMessageTransmitterV2(localMessageTransmitter).localDomain();
+    if (LOCAL_DOMAIN == DESTINATION_DOMAIN) revert InvalidLocalDomain();
+  }
+
+  /// @dev No use case to receive ETH; prevent accidental transfers.
+  receive() external payable {
+    if (msg.value > 0) revert CannotReceiveEther();
```
Comment (efecarranza) [inline src/bridges/cctp/CctpBridgeSteward.sol]: Whenever we've had something like this, Aave Labs have asked us to remove it hah.
We also have the rescueEth function so we could just use that if it ever happens that we have rescuable ETH

Comment (efecarranza) [reply src/bridges/cctp/CctpBridgeSteward.sol]: For example: https://github.com/aave-dao/gho-origin/pull/5/changes#diff-a6b551a2e12059c984f17705b48848ce7ecec1a99236117aa4817b92b90e8b2dR108

File: src/bridges/cctp/CctpBridgeSteward.sol (hunk 1 above)
Comment (efecarranza) [reply src/bridges/cctp/CctpBridgeSteward.sol]: Also, can be hardcoded to AaveV3Ethereum.COLLECTOR as constant

File: src/bridges/cctp/CctpBridgeSteward.sol [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  /// @inheritdoc ICctpBridgeSteward
+  function bridge(uint256 amount, uint256 maxFee, TransferSpeed speed) external onlyOwnerOrGuardian {
+    if (amount == 0) revert InvalidZeroAmount();
+    if (maxFee >= amount) revert InvalidMaxFee(maxFee, amount);
```
Comment (efecarranza) [inline src/bridges/cctp/CctpBridgeSteward.sol:67]: Is fee deducted from amount or has to be paid in addition?  Asking cause of the transfer in

File: tests/bridges/cctp/CctpBridgeSteward.t.sol [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+  function _addressToBytes32(address addr) internal pure returns (bytes32) {
+    return bytes32(uint256(uint160(addr)));
+  }
+
+  function _fundCollector(uint256 amount) internal {
+    deal(address(usdc), collector, amount);
+  }
+
+  function setUp() public virtual {
+    try vm.activeFork() returns (
```
Comment (efecarranza) [inline tests/bridges/cctp/CctpBridgeSteward.t.sol]: I personally think we should fork liberally when it's stuff that works with other protocols rather than being protocols themselves, and also I'm ok with just testing one flow like arbitrum to mainnet and one file. 
Don't see much value in the mocking.

File: src/bridges/cctp/CctpBridgeSteward.sol (hunk 3 above)
Comment (jubeira) [reply src/bridges/cctp/CctpBridgeSteward.sol:67]: It is deducted.

Here's the implementation of token messenger: https://etherscan.deth.net/address/0x555e272506c06e7e559d57418563742afe363ec8. See `_depositForBurn#L309`.

Technically if `maxFee >= amount` it should revert in the token messenger, but IMO it's a good practice to explicitly enforce bounds on every argument for an external function. The extra overhead is minimal, and it makes it easier for the reader to reason about the contract's edge cases.

File: tests/bridges/cctp/CctpBridgeSteward.t.sol (hunk 4 above)
Comment (jubeira) [reply tests/bridges/cctp/CctpBridgeSteward.t.sol]: Sounds good; I was on the fence with this one.

I wanted to cover as much as possible while keeping the test lean, and truth is you only need the fork for the `bridge` function.

But I agree that it might come up as over-engineering :) I'll simplify

File: src/bridges/cctp/CctpBridgeSteward.sol (hunk 1 above)
Comment (jubeira) [reply src/bridges/cctp/CctpBridgeSteward.sol]: Sounds good; I assume the collector is not expected to change anytime soon, and if it does it's still a small change.

File: src/bridges/cctp/CctpBridgeSteward.sol (hunk 2 above)
Comment (jubeira) [reply src/bridges/cctp/CctpBridgeSteward.sol]: Removed.

**Review by efecarranza — APPROVED** [review]
